src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsServerRouter/DiagnosticsServerRouter.cs

```diff
@@ -0,0 +1,841 @@
+                }
+                finally
+                {
+                    networkStream.Socket.Blocking = blockingState;
+                }
+            }
+            else
+            {
+                connected = false;
+            }
+
+            return connected;
+        }
+
+        protected async Task CheckStreamConnectionAsync(Stream stream, CancellationToken token)
+        {
+            while (!token.IsCancellationRequested)
+            {
+                // Check if tcp stream connection is still available.
+                if (!CheckStreamConnection(stream, token))
+                {
+                    throw new EndOfStreamException();
+                }
+
+                try
+                {
+                    await Task.Delay(TcpServerConnectTimeoutMs, token).ConfigureAwait(false);
+                }
+                catch { }
+            }
+        }
+
+        protected bool IsCompletedSuccessfully(Task t)
+        {
+#if NETCOREAPP2_0_OR_GREATER
+            return t.IsCompletedSuccessfully;
+#else
+            return t.IsCompleted && !t.IsCanceled && !t.IsFaulted;
+#endif
+        }
+    }
+
+    /// <summary>
+    /// This class connects IPC Server - TCP Server router instances.
+    /// Supports NamedPipes/UnixDomainSocket server and TCP/IP server.
+    /// </summary>
+    internal class IpcServerTcpServerRouter : TcpServerRouter, IIpcServerTransportCallbackInternal
+    {
+        readonly string _ipcServerPath;
+
+        IpcServerTransport _ipcServer;
+
+        public int IpcServerConnectTimeout { get; set; } = Timeout.Infinite;
+
+        public IpcServerTcpServerRouter(string ipcServer, string tcpServer, int runtimeTimeoutMs, ILogger logger)
+            : base(tcpServer, runtimeTimeoutMs, logger)
+        {
+            _ipcServerPath = ipcServer;
+            if (string.IsNullOrEmpty(_ipcServerPath))
+                _ipcServerPath = GetDefaultIpcServerPath();
+
+            _ipcServer = IpcServerTransport.Create(_ipcServerPath, IpcServerTransport.MaxAllowedConnections, false);
+            _tcpServer.TransportCallback = this;
+        }
+
+        public static string GetDefaultIpcServerPath()
+        {
+            int processId = Process.GetCurrentProcess().Id;
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                return Path.Combine(PidIpcEndpoint.IpcRootPath, $"dotnet-diagnostic-{processId}");
+            }
+            else
+            {
+                DateTime unixEpoch;
+#if NETCOREAPP2_1_OR_GREATER
+                unixEpoch = DateTime.UnixEpoch;
+#else
+                unixEpoch = new DateTime(1970, 1, 1, 0, 0, 0, 0, DateTimeKind.Utc);
+#endif
+                TimeSpan diff = Process.GetCurrentProcess().StartTime.ToUniversalTime() - unixEpoch;
+                return Path.Combine(PidIpcEndpoint.IpcRootPath, $"dotnet-diagnostic-{processId}-{(long)diff.TotalSeconds}-socket");
```

Comment:
Will this be consistent for CoreCLR and Mono? More specifically, does the `Sytem.Process.StartTime` property match the timestamp that the native code used to disambiguate the socket name?

Reply:
The most important thing is the format of the string and the you have a timestamp connected to a specific process ID relative to a fixed point (like unix epoch or system boot time).

When running on OSX, I get the name of uds created by native runtime in the same format as the one produced from above code.

Reply:
> When running on OSX, I get the name of uds created by native runtime in the same format as the one produced from above code.

If this is producing the same UDS name then the logic should work. If we find that the name is different on some platforms, we may want to reduce the resolution of the timestamp to allow for any drift between the timestamp reported by `System.Diagnostics.Process` and the timestamp the target app got when it created the UDS.